Classes/EventListener/CheckProductAvailability.php

```diff
@@ -10,86 +10,107 @@
  * For the full copyright and license information, please read the
  * LICENSE file that was distributed with this source code.
  */
+
+use Extcode\Cart\Domain\Model\Cart\BeVariant as CartProductBeVariant;
+use Extcode\Cart\Domain\Model\Cart\Product as CartProduct;
 use Extcode\Cart\Event\CheckProductAvailabilityEvent;
-use Extcode\CartProducts\Domain\Model\Product\Product;
-use Extcode\CartProducts\Domain\Repository\Product\ProductRepository;
+use Extcode\CartProducts\Domain\DoctrineRepository\Product\BeVariantRepository;
+use Extcode\CartProducts\Domain\DoctrineRepository\Product\ProductRepository;
 use TYPO3\CMS\Core\Messaging\FlashMessage;
 use TYPO3\CMS\Core\Type\ContextualFeedbackSeverity;
 use TYPO3\CMS\Core\Utility\GeneralUtility;
 use TYPO3\CMS\Extbase\Utility\LocalizationUtility;
 
 class CheckProductAvailability
 {
+    private CheckProductAvailabilityEvent $event;
+
     public function __construct(
-        private readonly ProductRepository $productRepository
+        private readonly ProductRepository $productRepository,
+        private readonly BeVariantRepository $beVariantRepository,
     ) {}
 
     public function __invoke(CheckProductAvailabilityEvent $event): void
     {
-        $cart = $event->getCart();
+        $this->event = $event;
         $cartProduct = $event->getProduct();
-        $quantity = $event->getQuantity();
-
-        $mode = $event->getMode();
 
-        if ($cartProduct->getProductType() !== 'CartProducts') {
+        if ($cartProduct->getProductType() !== 'CartProducts' || $cartProduct->isHandleStock() === false) {
             return;
         }
 
-        $querySettings = $this->productRepository->createQuery()->getQuerySettings();
-        $querySettings->setRespectStoragePage(false);
-        $this->productRepository->setDefaultQuerySettings($querySettings);
+        if ($cartProduct->isHandleStockInVariants()) {
+            foreach ($cartProduct->getBeVariants() as $cartProductBeVariant) {
+                $this->checkStockForBeVariant($cartProductBeVariant, $cartProduct);
+            }
+        } else {
+            $this->checkStockForProduct($cartProduct);
+        }
+    }
 
-        $product = $this->productRepository->findByIdentifier($cartProduct->getProductId());
+    private function checkStockForProduct(CartProduct $cartProduct): void
+    {
+        $cart = $this->event->getCart();
+        $mode = $this->event->getMode();
+
+        $quantity = $this->event->getQuantity();
+        // $quantity is an array in the event that the number of products with variants is updated in the shopping cart.
```

Comment:
My proposal:
```php
// $quantity is an array in the case that the amount of a product with variants
// is updated within the shopping cart.
```

I was confused by the word 'event' (because we are in an EventListener and my brain was desperately looking for the connection to the `Event`. Using 'product' (singular) and adding 'within' makes even me understanding the meaning :)

Reply:
I've changed the wording of the first sentence.